### PR TITLE
Rewrite presence a little bit

### PR DESCRIPTION
**Pull request by devsnek**
**Please describe the changes this PR makes and why it should be merged:**
Adds support for the richest of presence, switches to new presence caching

**Semantic versioning classification:**  
- [x] This PR changes the library's interface (methods or parameters added)
  - [x] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

File: src/stores/PresenceStore.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+const { Presence } = require('../structures/Presence');
+
+class ClientPresenceStore extends PresenceStore {
+  create(data) {
+    if (this.has(data.user.id)) {
+      this.get(data.user.id).patch(data);
+    } else {
+      this.set(data.user.id, new Presence(this.client, data));
+    }
+  }
```
Comment (hydrabolt) [inline src/stores/PresenceStore.js]: You also have to return the presence here (all the other stores do)

File: src/stores/ClientPresenceStore.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+      afk: false,
+      since: null,
+      activity: null,
+    });
+  }
+
+  async setClientPresence({ status, since, afk, activity }) {
+    const applicationID = activity.application ? activity.application.id || activity.application : null;
+    let assets = new Collection();
+    if (activity.assets && applicationID) {
```
Comment (SpaceEEC) [inline src/stores/ClientPresenceStore.js]: Same error with `activity` being undefined here.

File: src/stores/ClientPresenceStore.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+      try {
+        const a = await this.client.api.applications(applicationID).assets.get();
```
Comment (SpaceEEC) [inline src/stores/ClientPresenceStore.js]: This should be `this.client.api.oauth2.applications(applicationID).assets.get()` (missing `oauth2`).

File: src/stores/ClientPresenceStore.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        state: activity.state,
+        assets: activity.assets ? {
+          large_text: activity.assets.largeText,
+          small_text: activity.assets.smallText,
+          large_image: assets.get(activity.assets.largeImage) || activity.assets.largeImage,
+          small_image: assets.get(activity.assets.smallImage) || activity.assets.smallImage,
+        } : undefined,
+        application_id: applicationID || undefined,
+        secrets: activity.secrets || undefined,
+        instance: activity.instance || undefined,
```
Comment (SpaceEEC) [inline src/stores/ClientPresenceStore.js]: I assume, that omitting `instance` is the same as setting it to `false`?

File: src/stores/ClientPresenceStore.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        for (const asset of a) assets.set(asset.name, asset.id);
+      } catch (err) {} // eslint-disable-line no-empty
+    }
+
+    const packet = {
+      afk: afk != null ? afk : false, // eslint-disable-line eqeqeq
+      since: since != null ? since : null, // eslint-disable-line eqeqeq
+      status: status || this.clientPresence.status,
+      game: activity ? {
+        type: Constants.ActivityTypes.indexOf(activity.type || 'PLAYING') || activity.type || 0,
```
Comment (SpaceEEC) [inline src/stores/ClientPresenceStore.js]: You can no longer pass a numeric type here (except 0) as the indexOf returns `-1` in that case which is not falsy.

File: src/structures/ClientUser.js [hunk 6]
```diff
@@ -258,35 +214,31 @@ class ClientUser extends User {
...
   setStatus(status) {
     return this.setPresence({ status });
   }
 
   /**
-   * Sets the game the client user is playing.
-   * @param {?string} game Game being played
-   * @param {Object} [options] Options for setting the game
+   * Sets the activity the client user is playing.
+   * @param {?string} name activity being played
```
Comment (SpaceEEC) [inline src/structures/ClientUser.js]: Activity

File: src/structures/Presence.js [hunk 7]
```diff
@@ -43,46 +42,70 @@ class Presence {
...
       this.status === presence.status &&
-      this.game ? this.game.equals(presence.game) : !presence.game
+      this.activity ? this.activity.equals(presence.activity) : !presence.activity
     );
   }
 }
 
 /**
- * Represents a game that is part of a user's presence.
+ * Represents a activity that is part of a user's presence.
```
Comment (SpaceEEC) [inline src/structures/Presence.js]: an activity

File: src/structures/Presence.js [hunk 8]
```diff
@@ -91,5 +114,60 @@ class Game {
...
+  }
+
+  /**
+   * @param  {string} format Format of the image
+   * @param  {number} size Size of the iamge
+   * @returns {?string} small image url
+   */
+  smallImageURL({ format, size } = {}) {
+    if (!this.smallImage) return null;
+    return this.client.rest.cdn
```
Comment (SpaceEEC) [inline src/structures/Presence.js]: This class does not have a client property.

File: src/structures/Presence.js [hunk 9]
```diff
@@ -91,5 +114,60 @@ class Game {
...
+  }
+
+  /**
+   * @param  {string} format Format of the image
+   * @param  {number} size Size of the iamge
+   * @returns {?string} large image url
+   */
+  largeImageURL({ format, size } = {}) {
+    if (!this.largeImage) return null;
+    return this.client.rest.cdn
```
Comment (SpaceEEC) [inline src/structures/Presence.js]: Same here, no client property.

File: src/structures/ClientUser.js (hunk 6 above)
Comment (devsnek) [reply src/structures/ClientUser.js]: ?

Comment (SpaceEEC) [reply src/structures/ClientUser.js]: Uppercase A instead of lowercase.

File: src/structures/Presence.js [hunk 10]
```diff
@@ -92,6 +92,21 @@ class Activity {
     this.applicationID = data.application_id || null;
 
     /**
+     * Timestamps for the activity
+     * @type {Object}
```
Comment (SpaceEEC) [inline src/structures/Presence.js]: This is nullable.
And maybe add `@property` here to describe what it actually is.

File: src/structures/Presence.js [hunk 11]
```diff
@@ -92,6 +92,21 @@ class Activity {
...
+     */
+    this.timestamps = data.timestamps ? {
+      start: data.timestamps.start ? new Date(data.timestamps.start) : null,
+      end: data.timestamps.end ? new Date(data.timestamps.end) : null,
+    } : null;
+
+    /**
+     * Party of the activity
+     * @type {Object}
```
Comment (SpaceEEC) [inline src/structures/Presence.js]: Same as above.

File: src/stores/ClientPresenceStore.js [hunk 12]
```diff
@@ -40,6 +40,8 @@ class ClientPresenceStore extends PresenceStore {
           large_image: assets.get(activity.assets.largeImage) || activity.assets.largeImage,
           small_image: assets.get(activity.assets.smallImage) || activity.assets.smallImage,
         } : undefined,
+        timestamps: activity.timestamps || undefined,
```
Comment (SpaceEEC) [inline src/stores/ClientPresenceStore.js]: This should convert the Date object back to a unix timestamp.

Comment (devsnek) [reply src/stores/ClientPresenceStore.js]: not needed

**Pull request merged**